Source/DafnyCore/Compilers/CSharp/Compiler-Csharp.cs

```diff
@@ -265,27 +265,57 @@ public class CsharpCompiler : SinglePassCompiler {
 
           var wCtorParams = new ConcreteSyntaxTree();
           wCtorBody = wBody.Format($"public {name}({wCtorParams})").NewBlock();
+          EmitTypeDescriptorsForClass(typeParameters, cls, wTypeFields, wCtorParams, null, wCtorBody);
+        }
+      }
 
-          if (typeParameters != null) {
-            var sep = "";
-            foreach (var ta in TypeArgumentInstantiation.ListFromFormals(typeParameters)) {
-              if (NeedsTypeDescriptor(ta.Formal)) {
-                var fieldName = FormatTypeDescriptorVariable(ta.Formal.GetCompileName(Options));
-                var decl = $"{DafnyTypeDescriptor}<{TypeName(ta.Actual, wTypeFields, ta.Formal.tok)}> {fieldName}";
-                wTypeFields.WriteLine($"private {decl};");
-                if (ta.Formal.Parent == cls) {
-                  wCtorParams.Write($"{sep}{decl}");
-                }
+      return new ClassWriter(this, wBody, wCtorBody);
+    }
 
-                wCtorBody.WriteLine($"this.{fieldName} = {TypeDescriptor(ta.Actual, wCtorBody, ta.Formal.tok)};");
-                sep = ", ";
-              }
+    /// <summary>
```

Comment:
Adding constructor initialized fields to a class doesn't seem specific to C#. Can't this code be made more language agnostic?

Reply:
This would be great to rework for all the compilers. A much better architecture would be to rewrite the Dafny code into some Dafny intermediate form, and then each compiler would have a much simpler task. But I consider such a change a separate endeavor.